parsl/executors/radical/executor.py

```diff
@@ -0,0 +1,583 @@
+                    with open(p, 'wb') as ff:
+                        ff.write(r.content)
+                    file = File(p)
+
+                f = {'source': file.url,
+                     'action': rp.TRANSFER}
+                to_stage.append(f)
+
+            elif mode == 'out':
+                # this indicates that the user
+                # did not provided a specific
+                # output file and RP will stage out
+                # the task.output from pilot://task_folder
+                # to the CWD or file.url
+                if '/' not in file.url:
+                    f = {'source': file.filename,
+                         'target': file.url,
+                         'action': rp.TRANSFER}
+                    to_stage.append(f)
+            else:
+                raise ValueError('unknown staging mode')
+
+        return to_stage
+
+    def _bulk_collector(self):
+
+        bulk = list()
+
+        while True:
+
+            now = time.time()  # time of last submission
+
+            # collect tasks for min bulk time
+            # NOTE: total collect time could actually be max_time + min_time
+            while time.time() - now < self._max_bulk_time:
+
+                try:
+                    task = self._bulk_queue.get(block=True,
+                                                timeout=self._min_bulk_time)
+                except queue.Empty:
+                    task = None
+
+                if task:
+                    bulk.append(task)
+
+                if len(bulk) >= self._max_bulk_size:
+                    break
+
+            if bulk:
+                logger.debug('submit bulk: %d', len(bulk))
+                self.tmgr.submit_tasks(bulk)
+                bulk = list()
+
+    def submit(self, func, *args, **kwargs):
+        """
+        Submits tasks in stream mode or bulks (bulk mode)
+        to RADICAL-Pilot rp.TaskManager.
+        """
+        rp_tid = ru.generate_id('task.%(item_counter)06d', ru.ID_CUSTOM,
+                                ns=self.session.uid)
+        parsl_tid = int(rp_tid.split('task.')[1])
+
+        logger.debug("got Task {0} from parsl-dfk".format(parsl_tid))
+        task = self.task_translate(parsl_tid, func, args, kwargs)
+
+        # assign task id for rp task
+        task.uid = rp_tid
+
+        # set the future with corresponding id
+        self.future_tasks[rp_tid] = Future()
+
+        if self.bulk_mode:
+            # push task to rp submit thread
+            self._bulk_queue.put(task)
+        else:
+            # submit the task to rp
+            logger.debug("put {0} to rp-TMGR".format(rp_tid))
+            self.tmgr.submit_tasks(task)
+
+        return self.future_tasks[rp_tid]
+
+    def shutdown(self, hub=True, targets='all', block=False):
+        """Shutdown the executor, including all RADICAL-Pilot components."""
+        logger.info("RadicalPilotExecutor shutdown")
+        self.session.close(download=True)
+
+        return True
+
+    @property
```

Comment:
Since the start of development of this executor, there have been some changes to the scaling API: I think it's safe to remove all three of `scaling_enabled`, `scale_in` and `scale_out`: scaling_enabled was removed in PR #2545 because scaling behaviour now comes from subclassing `BlockProviderExecutor` and `scale_in`, `scale_out` are now only needed for `BlockProviderExecutor` subclasses - this work was intended to remove the need for all these stubs on executors which will manage their own workers (such as the radical pilot executor)